### ISSUE TITLE
Docker help isn't a helpful anymore w.r.t. optons

### ISSUE THREAD

**Issue by duglin** [docker/docker#18797]
Today's `docker build` include options in the help that look like this:
```
  --force-rm                      Always remove intermediate containers
  -m, --memory                    Memory limit
  -q, --quiet                     Suppress the build output and print image ID on success
  --rm=true                       Remove intermediate containers after a successful build
```
these used to be:
```
  --force-rm=false      Always remove intermediate containers
  -m, --memory=         Memory limit
  -q, --quiet=false     Suppress the verbose output generated by the containers
  --rm=true             Remove intermediate containers after a successful build
```

I may just be used to the old (2nd) format, but I don't find the new format as helpful.
First, its no longer clear to me which flags are booleans and which are not. In the past all booleans had either `=false` or `=true` in them.  Now just the ones that default to `true` have them. This means that a flag like `--quite` has no indication of whether its a boolean or not unless its VERY clear in the text - which is not always the case. For example, look at `--memory`.  To non-newbies it most likely means that we're supposed to supply the max memory, but how would a newbie know that its not a boolean flag indicating that whether Docker should try to "limit the memory used by the container", or not? Might not be the best example, but my point is that the old help text made it very clear (to non-english native folks) that the flag was a boolean.

Anyone else find the old text more helpful?  If its just me we can close this.

**Comment by thaJeztah** [comment]
This changed in https://github.com/docker/docker/pull/18082, and is in line with changes in the default Go flags package; https://go.googlesource.com/go/+/51d66601c45572484ed20f3a2be6f67d648b5f22

There's two things at hand here; don't show default "=false" for Boolean flags, and "non helpful descriptions"

I *don't* like the old way of documenting Boolean (false by default) flags. It's confusing, and I doubt many people will actually specify a value for Boolean flags. I've seen users being confused by the old output (I've used `--enable-something=false` as the instructions mention, but it doesn't work).

I *do* agree that, in our attempts to limit the length of the help-texts, they've become basically useless in many occasions (the "memory limit" is a good example). I think we should check the current help messages and improve their descriptions where needed.

This issue https://github.com/docker/docker/issues/18385 and PR https://github.com/docker/docker/pull/18387 may also be relevant here.

**Comment by duglin** [comment]
Just doing a quick check of something like the help for `ls`, I don't think the issue is the help text itself, but rather the way the options are presented.  I see things like `--format=WORD` and `--ignore=PATTERN`.  The presence of the word in caps is really helpful to me to know that more than just the flag is needed, and implicitly the lack of it means its a boolean flag.  In most (not all) of our options we nothing so there's no guide for users to know what's expected, so I'm wondering if we should follow other command's example here.

The more look at our stuff the more I'm wondering how we ended up with something so different from normal linux commands.

**Comment by thaJeztah** [comment]
Yes, I think that makes sense. One thing we should look at, is how standard Linux commands mention "default" values (especially, if they're not trivial). I do note that the help output there outputs multiple lines per flag/option (which we were still discussing in the other issue/PR)

For reference, this is a snippet of the `ls --help` output;

```
Mandatory arguments to long options are mandatory for short options too.
  -a, --all                  do not ignore entries starting with .
  -A, --almost-all           do not list implied . and ..
      --author               with -l, print the author of each file
  -b, --escape               print C-style escapes for nongraphic characters
      --block-size=SIZE      scale sizes by SIZE before printing them.  E.g.,
                               '--block-size=M' prints sizes in units of
                               1,048,576 bytes.  See SIZE format below.
  -B, --ignore-backups       do not list implied entries ending with ~
  -c                         with -lt: sort by, and show, ctime (time of last
                               modification of file status information)
                               with -l: show ctime and sort by name
                               otherwise: sort by ctime, newest first
  -C                         list entries by columns
      --color[=WHEN]         colorize the output.  WHEN defaults to 'always'
                               or can be 'never' or 'auto'.  More info below
  -d, --directory            list directory entries instead of contents,
                               and do not dereference symbolic links
  -D, --dired                generate output designed for Emacs' dired mode
  -f                         do not sort, enable -aU, disable -ls --color
  -F, --classify             append indicator (one of */=>@|) to entries
      --file-type            likewise, except do not append '*'
      --format=WORD          across -x, commas -m, horizontal -x, long -l,
                               single-column -1, verbose -l, vertical -C
```

**Comment by thaJeztah** [comment]
Oh, just noted that "defaults" are always mentioned in the *description*, similar to what's suggested in the changes to Golang flags

**Comment by duglin** [comment]
one other interesting thing to note is that I don't think "normal" commands allow true/false on boolean flags.  At least not with the ones I've tried.

**Comment by thaJeztah** [comment]
no, I don't think they do. I raised that issue a long time ago. OTOH, it allows us to have some flexibility; e.g. introduce a boolean flag and for a feature that will be deprecated (e.g. `--allow-old-feature=true`) and change the default later :wink:

**Comment by duglin** [comment]
yea but that nonconformity with other commands might be part of the reason people claim they're confused. I'd prefer if we followed linux's pattern instead of some new pattern that golang has come up with.

**Comment by duglin** [comment]
total bikeshedding here but the more I think about things like `--rm=true` the more odd to me it is. This really should have just been `--no-rm` so it could have a default value of "false".   

I wonder if we could deprecate all boolean flags that have a default value of `true` and replace them with their `--no-xxx` equivalent as a first step. Once we do that we can then get rid of the entire idea of specifying true/false at all.

We can also look at modifying the way we present non-boolean flags - IOW, add `=XXX` like normal commands have to indicate that a value is expected.

**Comment by thaJeztah** [comment]
Yup, see this discussion https://github.com/docker/docker/pull/5764 around that...

**Comment by thaJeztah** [comment]
Oh and this on Boolean flags accepting a value https://github.com/docker/docker/issues/5545#issuecomment-42885403

**Comment by duglin** [comment]
can't help but feel like we're reinventing a wheel, just a square one this time  :-)

**Comment by flx42** [comment]
Hello, maintainer of [nvidia-docker](https://github.com/NVIDIA/nvidia-docker) with @3XX0 here.
I agree that the changes introduced by #18082 made the output of `docker help` ambiguous. This is a problem for us since `nvidia-docker` is a wrapper script for the docker CLI that checks if the containers being launched need access to the NVIDIA GPUs. For this purpose, we must find the image on the command line and `inspect` it to see if it has a particular label (`LABEL com.nvidia.volumes.needed`). But finding where is the image name on the command-line requires being aware of the different flags and to know which of them are boolean (if an option is not boolean, you must skip the next argument too).
We used to rely on a simple regex after calling `docker help` to detect which flags are not boolean:
https://github.com/NVIDIA/nvidia-docker/blob/e7b792253f0d0ad1bed19a3236db0ac58aaf9563/tools/src/docker/docker.go#L51
This is not working anymore since Docker 1.10 because we have no way of knowing which flag is boolean when looking at the help:
```
 -i, --interactive               Keep STDIN open even if not attached
 -m, --memory                    Memory limit
```
They look the same, but one is boolean and the other is not. It used to be:
```
 -i, --interactive=false         Keep STDIN open even if not attached
  -m, --memory=                   Memory limit
```

This use case is certainly unusual but at the time we couldn’t find a better solution that could potentially work for all Docker versions. Now, the remaining options for us are a bit unpractical:
- Parse the man pages: the format used to describe options is more standard.
- Hardcode all the boolean options in a map, for all docker commands.
- Directly use the REST API of the Docker daemon.

Unless you are a planning a solution to allow plugins to inspect starting containers and inject arguments, it would be useful to improve the output of `docker help`; and as pointed out by @duglin, it is also currently confusing for new users.

Why don’t we use a volume plugin instead? We do, but we also need to take care of devices (/dev/nvidia*) and check if the image is compatible with the NVIDIA driver version on the host. That’s why we still need a full-fledged wrapper of the CLI.

Why don’t we use the Docker go package handling flags? We could, but given that `go get` always fetch the latest version, older versions of Docker would suddenly stop working for users.

**Comment by sean-jc** [comment]
#dibs

My thought is to implement a generic version of the 'expected input' information that is displayed in commands like `ls` by extending the `mflag.Value` interface to declare `Type() string`.  Type() would be used in PrintDefaults to show what type of input, if any, is expected by the flag.  Boolean flags that are false by default would not display their type.  Boolean flags that are true by default would show that they expect a BOOL since simply specifying the flag will have no effect.  Like `ls --help`, default value information would be relegated to the usage description.

In an ideal world, the expected input would be tailored to fit the flag instead of being a generic type, but doing so would require major modifications to the `mflag` API, as well as every flag definition.  At that point I think moving to a more feature rich CLI package would be a better use of time.


BEFORE

```
Options:

  --config=~/.docker              Location of client config files
  -D, --debug                     Enable debug mode
  -H, --host=[]                   Daemon socket(s) to connect to
  -h, --help                      Print usage
  -l, --log-level=info            Set the logging level
  --tls                           Use TLS; implied by --tlsverify
  --tlscacert=~/.docker/ca.pem    Trust certs signed only by this CA
  --tlscert=~/.docker/cert.pem    Path to TLS certificate file
  --tlskey=~/.docker/key.pem      Path to TLS key file
  --tlsverify                     Use TLS and verify the remote
  -v, --version                   Print version information and quit
```

AFTER
```
Options:

      --config=STRING       Location of client config files, default is '~/.docker'
  -D, --debug               Enable debug mode
  -H, --host=LIST           Daemon socket(s) to connect to
  -h, --help                Print usage
  -l, --log-level=STRING    Set the logging level, default is 'info'
      --tls                 Use TLS; implied by --tlsverify
      --tlscacert=STRING    Trust certs signed only by this CA, default is '~/.docker/ca.pem'
      --tlscert=STRING      Path to TLS certificate file, default is '~/.docker/cert.pem'
      --tlskey=STRING       Path to TLS key file, default is '~/.docker/key.pem'
      --tlsverify           Use TLS and verify the remote
  -v, --version             Print version information and quit
```

**Comment by thaJeztah** [comment]
Looks like XKCD discovered this issue;

![](http://imgs.xkcd.com/comics/man_page.png)

**Comment by thaJeztah** [comment]
@duglin do you think this issue should be closed now that we switched to Cobra?

**Comment by duglin** [comment]
@thaJeztah I hadn't really looked at the output of the new cobra stuff and its interesting. Took me a second to realize the default stuff is moved into the description text and the type of the arg is now shown in the flag definition - which is kind of neat.

However, looking at my original example (build), I see this:
```
--force-rm                Always remove intermediate containers
-m, --memory string       Memory limit
-q, --quiet               Suppress the build output and print image ID on success
--rm                      Remove intermediate containers after a successful build (default true)
```
Couple of things that jump out at me...
- its interesting that boolean flags don't show 'bool' in the left column, the way `-m` shows `string`. But ok, I guess Cobra thinks bools don't need it.
- but then what's odd is that `--rm` shows a default value of `true` while `--force-rm` doesn't show any default value at all. Same for `-q`.  Is this because a default bool value of `false`is just obvious? And something I mentioned in my original comment, how does the user know `-q` takes a bool or any arg at all? From what's shown above I would assume its a flag w/o an arg.  Normal linux bool flags don't take args from what I can tell - e.g. `ls -a` doesn't allow `ls -a=true`.
- Take `--rm` as well. No indication it takes an arg, except that we hint its default value is `true`. So a smart person might deduce that they can say `--rm false` but I think its also possible to think that `--rm` by itself would imply false since there's no indication I can pass an arg and no `--rm` means `true`, so `--rm` must toggle it to `false`, right?  :-)
- And, related to `--rm false` - I'm actually wrong in the previous bullet.  We're a bit inconsistent in how we allow values to be passed. For example, `--memory 2G` is ok, as is `--memory=2G`, but `--rm false` isn't.  While `--rm=false` is ok.  Bools *must* use `=` for some reason.

What's not clear to me is how much of this is due to Cobra vs go's flag parsing, because some of it is just weird when I compare it to normal Linux stuff (like the `ls -a` example I mentioned above).  I'd prefer if we moved to make things look more like Linux - and, in particular, bools didn't allow values. Which means the idea of `--xx=true` would never be allowed, it would always be just `--xx` and the default is always an implicit `--xx=false`.  But, would our parser allow that?

**Comment by thaJeztah** [comment]
> Normal linux bool flags don't take args from what I can tell - e.g. ls -a doesn't allow ls -a=true.

Boolean flags accepting values surprised me as well, so I'm happy that we don't mention them in the output, perhaps that's enough? Overall we should really try to never have them require a value (but I know we used it for some flags that we added for "migration" to a new value, e.g. `--userland-proxy=false`, which one day (hopefully) is changed to `--userland-proxy` to enable it.

> but then what's odd is that --rm shows a default value of true while --force-rm doesn't show any default value at all. 

I think this has to do with *where* the default is set; if a default is set in the *client*, then Cobra knows about it, and shows the default value. If, on the other hand, the default is set by the *daemon*, then it doesn't show.

From a "design" perspective, I think defaults should be set by the *daemon*, so that sending an "empty" API request leads to the *daemon* setting sensible defaults.

However, that leads to the CLI / help output not showing the defaults, so it's a challenge; possibly we need to "hardcode" the defaults in the help text description. Should be "ok" overall, unless the default changes, and you use a client version that doesn't match the daemon version. Perhaps we should just  accept that.
